src/vikwindow.c

```diff
@@ -1743,50 +1743,101 @@ static gboolean draw_scroll (VikWindow *vw, GdkEventScroll *event)
   return TRUE;
 }
 
-
-
 /********************************************************************************
  ** Ruler tool code
  ********************************************************************************/
-static void draw_ruler(VikViewport *vvp, GdkDrawable *d, GdkGC *gc, gint x1, gint y1, gint x2, gint y2, gdouble distance)
+static void draw_ruler(VikViewport *vvp, GdkDrawable *d, GdkGC *gc, const VikCoord *start, VikCoord *end)
 {
   PangoLayout *pl;
   gchar str[128];
   GdkGC *labgc = vik_viewport_new_gc ( vvp, "#cccccc", 1);
   GdkGC *thickgc = gdk_gc_new(d);
-  
-  gdouble len = sqrt((x1-x2)*(x1-x2) + (y1-y2)*(y1-y2));
-  gdouble dx = (x2-x1)/len*10; 
-  gdouble dy = (y2-y1)/len*10;
-  gdouble c = cos(DEG2RAD(15.0));
-  gdouble s = sin(DEG2RAD(15.0));
-  gdouble angle;
-  gdouble baseangle = 0;
+
+  gint x1, y1, x2, y2;
+  vik_viewport_coord_to_screen ( vvp, start, &x1, &y1 );
+
+  /* normalize end position in case the mouse pointer was moved out of the map */
+  if (end->mode == VIK_COORD_LATLON) {
+    end->east_west = fmod(end->east_west + 360, 360);
+    if (end->east_west > 180) {
+      end->east_west -= 360;
+    }
+  }
+  vik_viewport_coord_to_screen ( vvp, end, &x2, &y2 );
+
+
+  gdouble len, dx, dy, c, s, angle, angle_end, display_angle, baseangle;
   gint i;
 
-  /* draw line with arrow ends */
-  {
-    gint tmp_x1=x1, tmp_y1=y1, tmp_x2=x2, tmp_y2=y2;
-    a_viewport_clip_line(&tmp_x1, &tmp_y1, &tmp_x2, &tmp_y2);
-    gdk_draw_line(d, gc, tmp_x1, tmp_y1, tmp_x2, tmp_y2);
+  vik_viewport_compute_bearing ( vvp, x1, y1, x2, y2, &display_angle, &baseangle );
+  angle = fmod(baseangle  + DEG2RAD(vik_coord_angle ( start, end )), 2*M_PI);
+  angle_end = fmod(baseangle + DEG2RAD(vik_coord_angle_end ( start, end )), 2*M_PI);
+
+  gdouble distance = vik_coord_diff (start, end);
+
+  len = sqrt((x1-x2)*(x1-x2) + (y1-y2)*(y1-y2));
+  dx = (x2-x1)/len*10;
+  dy = (y2-y1)/len*10;
+  c = cos(DEG2RAD(15.0));
+  s = sin(DEG2RAD(15.0));
+
+  /* if the distance is less than 10km, the curvature definetely won't be visible */
+  if (distance < 10e3) {
+    /* draw line with arrow ends */
+    a_viewport_clip_line(&x1, &y1, &x2, &y2);
+    gdk_draw_line(d, gc, x1, y1, x2, y2);
+
+    /* orthogonal bars */
+    gdk_draw_line(d, gc, x1 - dy, y1 + dx, x1 + dy, y1 - dx);
+    gdk_draw_line(d, gc, x2 - dy, y2 + dx, x2 + dy, y2 - dx);
+    /* arrow components */
+    gdk_draw_line(d, gc, x2, y2, x2 - (dx * c + dy * s), y2 - (dy * c - dx * s));
+    gdk_draw_line(d, gc, x2, y2, x2 - (dx * c - dy * s), y2 - (dy * c + dx * s));
+    gdk_draw_line(d, gc, x1, y1, x1 + (dx * c + dy * s), y1 + (dy * c - dx * s));
+    gdk_draw_line(d, gc, x1, y1, x1 + (dx * c - dy * s), y1 + (dy * c + dx * s));
   }
 
-  a_viewport_clip_line(&x1, &y1, &x2, &y2);
-  gdk_draw_line(d, gc, x1, y1, x2, y2);
+  else {
+    gint last_x = x1;
+    gint last_y = y1;
+    gint x, y;
+
+    /* draw geodesic */
+    for (gdouble step=0;step<=1;step+=0.001) {
```

Comment:
A thousand steps might be a bit excessive. Seems fine on my computer, but maybe on a old computer this might be slow. Using a hundred steps (i.e. a value of 0.01), still seems to generate very reasonable 'curves', so would be a better trade off.
I don't think we'd need to bother with allowing configuration of this value (or guestimating the speed of the computer and trying to auto generate an appropriate value).